### PR TITLE
test/ unit tests for get Impersonation requests

### PR DESCRIPTION
**Pull request by Suvidh-kaushik**
Date: 12/06/2025

Developer Name: Suvidh Kaushik

---

## Issue Ticket Number

- Closes: #2200 

## Description

- This PR contains the unit tests for the model, validator of the get Impersonation requests logic
- **This PR must be merged after the Feature PR - #2442 and create route #2439 ** 
- **The import errors and build fail is because the feature PR has not been merged yet** 

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

N/A

</details>

## Test Coverage
<details>
<summary>Screenshots</summary>

Screenshot 2025-06-25 014155

Screenshot 2025-06-12 011652

Screenshot 2025-06-12 011723

Screenshot 2025-06-12 011746

</details>

## Additional Notes

Design Doc - [LINK](https://docs.google.com/document/d/1Qt0-knoFEWDl6Gdxr0W1xhPy87H-ixwj9Y55Q0bKMUo/edit?tab=t.0#heading=h.w1nbw3goorza)

PRD - [LINK](https://docs.google.com/document/d/1KI7zp81vevaIX3nDNgTsCHDY1iWmad-XFaVgEi3nH2Q/edit?tab=t.0)

### What change is being made?
Add unit tests for `getImpersonationRequestsValidator` middleware and `impersonationRequests` model functionalities to ensure the validators and model methods function correctly.

### Why are these changes being made?
These changes are being made to enhance the test coverage and ensure the correctness of the validation and retrieval logic for impersonation requests, thereby improving the reliability of the application. This implementation simulates various scenarios, ensuring the middleware and models handle both valid and invalid inputs and edge cases appropriately.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`

### CODE REVIEW

File: test/unit/models/impersonationRequests.test.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
+import { expect } from "chai";
+import cleanDb from "../../utils/cleanDb";
+import * as impersonationModel from "../../../models/impersonationRequests";
+import {
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:4]: this branch will be merged after the feature PR has been merged hence this error, once the feat PR is merged an i update the branch we will not get this error

File: test/unit/models/impersonationRequests.test.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  impersonationRequestsBodyData,
+  updateImpersonationRequestApproved,
+  updateImpersonationRequestRejected,
+} from "../../fixtures/impersonation-requests/impersonationRequests";
+import { REQUEST_STATE } from "../../../constants/requests";
+import addUser from "../../utils/addUser";
+import userDataFixture from "./../../fixtures/user/user";
+import sinon from "sinon";
+import { CreateImpersonationRequestModelDto } from "../../../types/impersonationRequest";
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:13]: this branch will be merged after the feature PR has been merged hence this error, once the feat PR is merged an i update the branch we will not get this error, i am upating the type name in that feat PR

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+const { expect } = chai;
+
+describe("Impersonation Request Validators", function () {
+  let req: any;
+  let res: any;
+  let requestBody: CreateImpersonationRequestBody = {
+    impersonatedUserId: "randomId",
+    reason: "Testing purpose"
+  };
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/middlewares/impersonationRequests.test.ts]: fixed it

File: test/unit/models/impersonationRequests.test.ts [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+const userData = userDataFixture();
+let testUserId: string;
+let requestBody: CreateImpersonationRequestModelDto;
+let impersonationRequests = [];
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts]: fixed

File: test/unit/models/impersonationRequests.test.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    const userIdPromises = [addUser(userData[16])];
+    const [userId] = await Promise.all(userIdPromises);
+    testUserId = userId;
+    requestBody = impersonationRequestsBodyData[0];
+  });
+
+  afterEach(async () => {
+    sinon.restore();
+    await cleanDb();
+  });
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts]: fixed

File: test/unit/models/impersonationRequests.test.ts [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+      expect(result.allRequests.length).to.be.equal(2);
+      expect(result.page).to.be.equal(2);
+    });
+
+    it("Should return empty array if no data is found", async () => {
+      await cleanDb();
+      const query = { status: REQUEST_STATE.PENDING };
+      const impersonationRequestData = await impersonationModel.getImpersonationRequests(query);
+      expect(impersonationRequestData).to.be.equal(null);
+    });
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:138]: fixed

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  let nextSpy: sinon.SinonSpy;
+
+  beforeEach(function () {
+    res = {
+      boom: {
+        badRequest: sinon.spy(),
+      },
+    };
+    nextSpy = sinon.spy();
+  });
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/middlewares/impersonationRequests.test.ts:19]: this is causing tests to fail keeping it without a type helps me when i want to reuse this req and res for using it in other functionalities tests in same file like update and create

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import chai from "chai";
+import sinon from "sinon";
+import { getImpersonationRequestsValidator } from "../../../middlewares/validators/impersonationRequests";
+
+const { expect } = chai;
+
+describe("Impersonation Request Validators", function () {
+  let req: any;
+  let res: any;
```
Comment (Achintya-Chatterjee) [inline test/unit/middlewares/impersonationRequests.test.ts:9]: use proper types instead of any

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    };
+    nextSpy = sinon.spy();
+  });
+
+ describe("getImpersonationRequestsValidator", function () {
+    it("should validate for a valid get impersonation requests query", async function () {
+      req = {
+        query: {
+          dev: true,
+          status: "APPROVED",
```
Comment (Achintya-Chatterjee) [inline test/unit/middlewares/impersonationRequests.test.ts:26]: We should also consider testing with a subset of the query, like only `dev=true`, to ensure partial queries don’t break the validator.

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+      await getImpersonationRequestsValidator(req as GetImpersonationControllerRequest, res as ImpersonationRequestResponse, nextSpy);
+      expect(nextSpy.calledOnce).to.be.true;
+    });
+
+    it("should not validate for an invalid status in query", async function () {
+      req = {
+        query: {
+          dev: true,
+          status: "INVALID_STATUS",
+        },
```
Comment (Achintya-Chatterjee) [inline test/unit/middlewares/impersonationRequests.test.ts:38]: Also add a test for **missing required fields** (like if `status` is missing when required)

File: test/unit/models/impersonationRequests.test.ts [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+      expect(impersonationRequestData.page).to.be.equal(2);
+    });
+
+    it("should return the next page of results using next cursor", async () => {
+      const firstPage = await impersonationModel.getImpersonationRequests({ size: 2 });
+      expect(firstPage.next).to.exist;
+      const nextPage = await impersonationModel.getImpersonationRequests({ size: 2, next: firstPage.next });
+      expect(nextPage.allRequests.length).to.be.at.most(2);
+      expect(nextPage.allRequests[0].id).to.not.equal(firstPage.allRequests[0].id);
+    });
```
Comment (Achintya-Chatterjee) [inline test/unit/models/impersonationRequests.test.ts:162]: Add an explicit assertion on `nextPage.page` or cursor presence to help validate pagination logic further.

**Review by Achintya-Chatterjee — CHANGES_REQUESTED** [review]

File: test/unit/middlewares/impersonationRequests.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/middlewares/impersonationRequests.test.ts:23]: I have updated this

File: test/unit/middlewares/impersonationRequests.test.ts (hunk 10 above)
Comment (Suvidh-kaushik) [reply test/unit/middlewares/impersonationRequests.test.ts]: i do not think this will be needed since get route has not required fields all fields are optional

File: test/unit/models/impersonationRequests.test.ts (hunk 11 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts]: have added this and am testing if the page has next and prev present along with page as null since we are using curosr based pagination by default

File: test/unit/middlewares/impersonationRequests.test.ts (hunk 10 above)
Comment (Achintya-Chatterjee) [reply test/unit/middlewares/impersonationRequests.test.ts]: okay

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    };
+    nextSpy = sinon.spy();
+  });
+
+  describe("getImpersonationRequestsValidator", function () {
+    it("should validate for a valid get impersonation requests query", async function () {
+      req = {
+        query: {
+          dev: "true",
+          status: "APPROVED",
```
Comment (shubhdevelop) [inline test/unit/middlewares/impersonationRequests.test.ts:30]: validate for other query params like createdBy createdFor , request id etc

File: test/unit/models/impersonationRequests.test.ts [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+  });
+
+  describe("getImpersonationRequests", () => {
+    beforeEach(async () => {
+      impersonationRequests = [];
+      impersonationRequests = await Promise.all(
+        impersonationRequestsBodyData.slice(0, 5).map((data) =>
+          impersonationModel.createImpersonationRequest(data)
+        )
+      );
```
Comment (shubhdevelop) [inline test/unit/models/impersonationRequests.test.ts:52]: can be a single statement no need to first intialise with empty array

File: test/unit/middlewares/impersonationRequests.test.ts (hunk 12 above)
Comment (Suvidh-kaushik) [reply test/unit/middlewares/impersonationRequests.test.ts:30]: not required according to me since if one is passing the validation others will also pass and since the params are optional we do not need to test all

File: test/unit/models/impersonationRequests.test.ts (hunk 13 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:52]: reason i am intializing with empty array is becuase in later update tests i am approving some requests or rejecting some requests data and without making the array empty if i try to push it will give unwanted results

File: test/unit/middlewares/impersonationRequests.test.ts [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        },
+      };
+      await getImpersonationRequestsValidator(
+        req as any,
```
Comment (MayankBansal12) [inline test/unit/middlewares/impersonationRequests.test.ts:34]: let's replace this type here and from line 48

File: test/unit/models/impersonationRequests.test.ts (hunk 13 above)
Comment (shubhdevelop) [reply test/unit/models/impersonationRequests.test.ts:52]: before every executioni of `it` the imperstationRequest would be filled with same data? that what beforeEach does

Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:52]: yes before every execution if i do not make it empty it will keep pushing into array

File: test/unit/middlewares/impersonationRequests.test.ts (hunk 14 above)
Comment (Suvidh-kaushik) [reply test/unit/middlewares/impersonationRequests.test.ts:90]: fixed this

**Review by Achintya-Chatterjee — DISMISSED** [review]
